### ISSUE TITLE
Enhancement suggestion: Interpolation API

### ISSUE THREAD

**Issue by mozfreddyb** [cure53/DOMPurify#46]
TLDR: Secure interpolation API in DOMPurify.
Usage:
```js
interpolate`<a href='${href}' title='${title}'>${text}</a>`
```
Returns a safe escaping of the variables href, title and text *according to their context*.

Implementation:
Using [ES6 template strings](http://tc39wiki.calculist.org/es6/template-strings/#) (and possibly a transpiler to make this work in non-ES6 environments), the interpolation function gets the context in which the variables are used. The function can then construct a safe clone of the template, e.g. `<a href='1' title='2'>3</a>`. When doing this, a hash table of identifiers and variable names is built (`1: href, 2: title, 3: text`).
This template may be safely inspected without any DOM clobbering concerns (Note: the thread model is that the template is safe and the variables are not).

Parsing & walking this template should be done in the typical DOMPurify style and pause whenever a placeholder value is seen (this can be achieved by looking at the hash table). The parser is stateful and DOM-aware, so it  knows the current context in which the placeholder occurs and escapes accordingly. Contexts-aware escaping notes that combinations of attribute and tag need their own sanitizing function, e.g. URL escaping for `href`, disallowing `javascript:` etc.

When the template DOM has been completely parsed and each placeholder value has been replaced with an escaped input value, it's being transformed back into a string and returned.

**Comment by mozfreddyb** [comment]
Mario suggested doing this in a branch or exploring this using the hooking API, before taking it into DOMPurify.

**Comment by cure53** [comment]
I like the idea but I don't see it arriving in the core anytime soon. However a branch and maybe a plugin or even a split project would be cool.

I am wondering though - how would you image the API to look like?

**Comment by mozfreddyb** [comment]
```js
// let's assume these three vars are untrusted
var href = "javascript:alert(1)";
var title = "click me";
var text = "innocent text <script>alert(1)</script>";

whatever.innerHTML = interpolate`<a href='${href}' title='${title}'>${text}</a>`
// returns <a href='' title='click me'>innocent text &lt;script&gt;alert(1)&lt;/script&gt;</a>
```

**Comment by cure53** [comment]
Would you expect DOMPurify to sanitize the single components or rather the final result? 
I'd assume the latter, just to be sure.

**Comment by mozfreddyb** [comment]
I have assumed to do the context-aware encoding in DOMPurify, but that's beyond its capabilities and it dawns me that there is very little overlap between my interpolation suggestion and DOMPurify.

Thanks for this good question :-) This will mean my interpolation should really live outside of DOMPurify or use it (at the end) to sanitize the final result..

**Comment by cure53** [comment]
@mozfreddyb Should we close this for now?

**Comment by mozfreddyb** [comment]
aye
